test-cloud/test-cloud-remote/src/test/java/org/kie/cloud/integrationtests/architecture/AbstractCloudArchitectureIntegrationTest.java

```diff
@@ -0,0 +1,188 @@
+import org.kie.server.integrationtests.router.client.KieServerRouterClient;
+import org.kie.server.router.Configuration;
+
+public abstract class AbstractCloudArchitectureIntegrationTest extends AbstractCloudIntegrationTest<GenericScenario> {
+
+    protected static final String SMART_ROUTER_ID = "test-kie-router";
+
+    protected static final String
+            CONTAINER_ID_ABC = CONTAINER_ID + "-abc",
+            CONTAINER_ID_DEF = CONTAINER_ID + "-def",
+            CONTAINER_ID_GHI = CONTAINER_ID + "-ghi";
+    protected static final String
+            KIE_CONTAINER_DEPLOYMENT_ABC = CONTAINER_ID_ABC + "=" + PROJECT_GROUP_ID + ":" + DEFINITION_PROJECT_NAME + ":" + DEFINITION_PROJECT_VERSION,
+            KIE_CONTAINER_DEPLOYMENT_DEF = CONTAINER_ID_DEF + "=" + PROJECT_GROUP_ID + ":" + DEFINITION_PROJECT_NAME + ":" + DEFINITION_PROJECT_VERSION,
+            KIE_CONTAINER_DEPLOYMENT_GHI = CONTAINER_ID_GHI + "=" + PROJECT_GROUP_ID + ":" + DEFINITION_PROJECT_NAME + ":" + DEFINITION_PROJECT_VERSION;
+
+    protected static final String REPO_BRANCH = "master";
+
+    protected static final String CONTROLLER_NAME = "controller";
+    protected static final String SMART_ROUTER_NAME = "smart-router";
+
+    protected static final String
+            KIE_SERVER_ABC_NAME = "kie-server-abc",
+            KIE_SERVER_DEF_NAME = "kie-server-def",
+            KIE_SERVER_GHI_NAME = "kie-server-ghi";
+
+    protected static final String PROJECT_SOURCE_FOLDER = "/kjars-sources";
+
+    protected static final String PORT = "80";
+
+    protected String repositoryName;
+
+    protected String getKieServerId(KieServicesClient kieServerClient) {
+        ServiceResponse<KieServerInfo> serverInfo = kieServerClient.getServerInfo();
+        assertThat(serverInfo.getType()).isEqualTo(ServiceResponse.ResponseType.SUCCESS);
+
+        return serverInfo.getResult().getServerId();
+    }
+
+    protected void verifyClientKieServersLocationCount(KieServicesClient kieServerClient, final int numberOfExpectedServers) {
+        if (numberOfExpectedServers == 0) {
+            ServiceResponse<KieServerInfo> serverInfo;
+            try {
+                serverInfo = kieServerClient.getServerInfo();
+            } catch (KieServicesHttpException ex) {
+                //ok
+                assertThat(ex.getHttpCode()).isEqualTo(503);
+                return;
+            }
+            assertThat(serverInfo.getType()).isEqualTo(ServiceResponse.ResponseType.FAILURE);
+            assertThat(serverInfo.getMsg()).contains("404");
+        } else {
+            Set<String> locations = new HashSet<>();
+            for (int i = 0; i < numberOfExpectedServers * 3; i++) {
+                ServiceResponse<KieServerInfo> serverInfo = kieServerClient.getServerInfo();
+                assertThat(serverInfo.getType()).isEqualTo(ServiceResponse.ResponseType.SUCCESS);
+
+                locations.add(serverInfo.getResult().getLocation());
```

Comment:
Another good case for `SoftAssertions`. Let it all evaluate, report all at the same time.

Reply:
I don't think so. I need to know here, if Service respond was success, so then I can get service response result without  NPE.

Reply:
So the result of the method depends on whether or not the assertion fails? Sorry, that's borderline evil. :-)

My suggestion: do whatever you need in the method, return the locations, and then assert on them.